pkg/stanza/fileconsumer/header.go

```diff
@@ -0,0 +1,274 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fileconsumer // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer"
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+
+	"go.opentelemetry.io/collector/extension/experimental/storage"
+	"go.uber.org/zap"
+	"golang.org/x/text/encoding"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/entry"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator/helper"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/pipeline"
+)
+
+type HeaderConfig struct {
+	Pattern           string            `mapstructure:"pattern"`
+	MetadataOperators []operator.Config `mapstructure:"metadata_operators"`
+
+	// these are set by the "build" function
+	matchRegex *regexp.Regexp
+	splitFunc  bufio.SplitFunc
+}
+
+// validate returns an error describing why the configuration is invalid, or nil if the configuration is valid.
+func (hc *HeaderConfig) validate() error {
+	_, err := regexp.Compile(hc.Pattern)
+	if err != nil {
+		return fmt.Errorf("invalid `pattern`: %w", err)
+	}
+
+	if len(hc.MetadataOperators) == 0 {
+		return errors.New("at least one operator must be specified for `metadata_operators`")
+	}
+
+	nopLogger := zap.NewNop().Sugar()
+	outOp := newHeaderPipelineOutput(nopLogger)
+	p, err := pipeline.Config{
+		Operators:     hc.MetadataOperators,
+		DefaultOutput: outOp,
+	}.Build(nopLogger)
+
+	if err != nil {
+		return fmt.Errorf("failed to build pipelines: %w", err)
+	}
+
+	if !p.Operators()[0].CanProcess() {
+		return errors.New("first operator must be able to process entries")
+	}
```

Comment:
Hmm, this is a good point. We don't want this pipeline to have any other input. I think we should actually validate that all of the operators `CanProcess` because this excludes input operators.

Reply:
Good idea.

Some other validations we could maybe do?:
1. Disallow the `filter` processor; If a log is filtered, the header won't ever receive the entry it expects in its output.
2. Ban output operators (op.CanOutput() == false) - same reasoning as above.

Are we OK with doing those validation too?

Reply:
Yeah, these make sense as well.